File: ai-code-frontend/src/services/backend/userController.ts
import request from '@/utils/request';
import type {
  UserLoginParams,
  UserRegisterParams,
  UserUpdateParams,
  UserUpdateInfoParams,
  UserUpdatePasswordParams,
  UserUpdateAvatarParams,
  SendEmailCodeParams,
  UserQueryParams,
  DeleteParams,
  UserAddParams,
} from '@/api/params/userParams';
import type {
  UserVO,
  UserDetailVO,
  UserLoginVO,
  UserRegisterVO,
  UserPointsVO,
  UserSimpleVO,
} from '@/api/vo/userVO';
import type { PageResponseVO } from '@/api/vo';

/**
 * 添加用户
 */
export async function addUser(body: UserAddParams) {
  return request.post<UserAddParams, PageResponseVO<number>>('/user/add', body);
}

/**
 * 删除用户
 */
export async function deleteUser(body: DeleteParams) {
  return request.post<DeleteParams, PageResponseVO<boolean>>('/user/delete', body);
}

/**
 * 根据ID获取用户
 */
export async function getUserById(id: string) {
  return request.get<string, PageResponseVO<UserVO>>('/user/get', { params: { id } });
}

/**
 * 获取当前登录用户
 */
export async function getLoginUser() {
  return request.get<void, PageResponseVO<UserLoginVO>>('/user/get/login');
}

/**
 * 根据ID获取用户VO
 */
export async function getUserVoById(id: string) {
  return request.get<string, PageResponseVO<UserVO>>('/user/get/vo', { params: { id } });
}

/**
 * 获取用户信息
 */
export async function getUserInfo(id: string) {
  return request.get<string, PageResponseVO<UserVO>>('/user/get/vo', { params: { id } });
}

/**
 * 分页查询用户列表
 */
export async function listUserVoByPage(body: UserQueryParams) {
  return request.post<UserQueryParams, PageResponseVO<PageResponseVO<UserVO>>>('/user/list/page/vo', body);
}

/**
 * 用户登录
 */
export async function userLogin(body: UserLoginParams) {
  return request.post<UserLoginParams, PageResponseVO<UserLoginVO>>('/user/login', body);
}

/**
 * 用户登出
 */
export async function userLogout() {
  return request.post<void, PageResponseVO<boolean>>('/user/logout', {});
}

/**
 * 用户注册
 */
export async function userRegister(body: UserRegisterParams) {
  return request.post<UserRegisterParams, PageResponseVO<number>>('/user/register', body);
}

/**
 * 更新用户
 */
export async function updateUser(body: UserUpdateParams) {
  return request.post<UserUpdateParams, PageResponseVO<boolean>>('/user/update', body);
}

/**
 * 更新用户信息
 */
export async function updateUserInfo(body: UserUpdateInfoParams) {
  return request.post<UserUpdateInfoParams, PageResponseVO<boolean>>('/user/update/info', body);
}

/**
 * 更新用户密码
 */
export async function updateUserPassword(body: UserUpdatePasswordParams) {
  return request.post<UserUpdatePasswordParams, PageResponseVO<boolean>>('/user/update/password', body);
}

/**
 * 更新用户头像
 */
export async function updateUserAvatar(body: UserUpdateAvatarParams) {
  return request.post<UserUpdateAvatarParams, PageResponseVO<boolean>>('/user/update/avatar', body);
}

/**
 * 发送邮箱验证码
 */
export async function sendEmailCode(body: SendEmailCodeParams) {
  return request.post<SendEmailCodeParams, PageResponseVO<boolean>>('/user/email/send', body);
}

/**
 * 通过邮箱注册
 */
export async function userRegisterByEmail(body: UserRegisterParams) {
  return request.post<UserRegisterParams, PageResponseVO<number>>('/user/register/email', body);
}

/**
 * 获取用户积分
 */
export async function getUserPoint(userId: string) {
  return request.get<string, PageResponseVO<UserPointsVO>>('/userPoint/get', { params: { userId } });
}

File: src/main/java/com/mashang/aicode/web/controller/UserController.java
package com.mashang.aicode.web.controller;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.util.StrUtil;
import com.mashang.aicode.web.annotation.AuthCheck;
import com.mashang.aicode.web.common.BaseResponse;
import com.mashang.aicode.web.common.DeleteRequest;
import com.mashang.aicode.web.common.ResultUtils;
import com.mashang.aicode.web.constant.UserConstant;
import com.mashang.aicode.web.exception.BusinessException;
import com.mashang.aicode.web.exception.ErrorCode;
import com.mashang.aicode.web.exception.ThrowUtils;
import com.mashang.aicode.web.model.dto.user.*;
import com.mashang.aicode.web.model.entity.User;
import com.mashang.aicode.web.model.vo.LoginUserVO;
import com.mashang.aicode.web.model.vo.UserVO;
import com.mashang.aicode.web.service.UserService;
import com.mybatisflex.core.paginate.Page;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * 用户 控制层。
 */
@RestController
@RequestMapping("/user")
public class UserController {

    @Resource
    private UserService userService;

    /**
     * 用户注册
     *
     * @param userRegisterRequest 用户注册请求
     * @return 注册结果
     */
    @PostMapping("/register")
    public BaseResponse<Long> userRegister(@RequestBody UserRegisterRequest userRegisterRequest) {
        ThrowUtils.throwIf(userRegisterRequest == null, ErrorCode.PARAMS_ERROR);
        String userAccount = userRegisterRequest.getUserAccount();
        String userPassword = userRegisterRequest.getUserPassword();
        String checkPassword = userRegisterRequest.getCheckPassword();
        long result = userService.userRegister(userAccount, userPassword, checkPassword);
        return ResultUtils.success(result);
    }

    /**
     * 用户登录
     *
     * @param userLoginRequest 用户登录请求
     * @param request          请求对象
     * @return 脱敏后的用户登录信息
     */
    @PostMapping("/login")
    public BaseResponse<LoginUserVO> userLogin(@RequestBody UserLoginRequest userLoginRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(userLoginRequest == null, ErrorCode.PARAMS_ERROR);
        String userAccount = userLoginRequest.getUserAccount();
        String userPassword = userLoginRequest.getUserPassword();
        LoginUserVO loginUserVO = userService.userLogin(userAccount, userPassword, request);
        
        // 获取 token 并返回
        String token = cn.dev33.satoken.stp.StpUtil.getTokenValue();
        loginUserVO.setToken(token);
        
        return ResultUtils.success(loginUserVO);
    }

    @GetMapping("/get/login")
    public BaseResponse<LoginUserVO> getLoginUser(HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        return ResultUtils.success(userService.getLoginUserVO(loginUser));
    }

    /**
     * 用户注销
     *
     * @param request 请求对象
     * @return
     */
    @PostMapping("/logout")
    public BaseResponse<Boolean> userLogout(HttpServletRequest request) {
        ThrowUtils.throwIf(request == null, ErrorCode.PARAMS_ERROR);
        boolean result = userService.userLogout(request);
        return ResultUtils.success(result);
    }

    /**
     * 创建用户
     */
    @PostMapping("/add")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Long> addUser(@RequestBody UserAddRequest userAddRequest) {
        ThrowUtils.throwIf(userAddRequest == null, ErrorCode.PARAMS_ERROR);
        User user = new User();
        BeanUtil.copyProperties(userAddRequest, user);
        // 默认密码 12345678
        final String DEFAULT_PASSWORD = "12345678";
        String encryptPassword = userService.getEncryptPassword(DEFAULT_PASSWORD);
        user.setUserPassword(encryptPassword);
        boolean result = userService.save(user);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(user.getId());
    }

    /**
     * 根据 id 获取用户（仅管理员）
     */
    @GetMapping("/get")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<User> getUserById(long id) {
        ThrowUtils.throwIf(id <= 0, ErrorCode.PARAMS_ERROR);
        User user = userService.getById(id);
        ThrowUtils.throwIf(user == null, ErrorCode.NOT_FOUND_ERROR);
        return ResultUtils.success(user);
    }

    /**
     * 根据 id 获取包装类
     */
    @GetMapping("/get/vo")
    public BaseResponse<UserVO> getUserVOById(long id) {
        BaseResponse<User> response = getUserById(id);
        User user = response.getData();
        return ResultUtils.success(userService.getUserVO(user));
    }

    /**
     * 删除用户
     */
    @PostMapping("/delete")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> deleteUser(@RequestBody DeleteRequest deleteRequest) {
        if (deleteRequest == null || deleteRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        boolean b = userService.removeById(deleteRequest.getId());
        return ResultUtils.success(b);
    }

    /**
     * 更新用户
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> updateUser(@RequestBody UserUpdateRequest userUpdateRequest) {
        if (userUpdateRequest == null || userUpdateRequest.getId() == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User user = new User();
        BeanUtil.copyProperties(userUpdateRequest, user);
        boolean result = userService.updateById(user);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(true);
    }

    /**
     * 分页获取用户封装列表（仅管理员）
     *
     * @param userQueryRequest 查询请求参数
     */
    @PostMapping("/list/page/vo")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Page<UserVO>> listUserVOByPage(@RequestBody UserQueryRequest userQueryRequest) {
        ThrowUtils.throwIf(userQueryRequest == null, ErrorCode.PARAMS_ERROR);
        long pageNum = userQueryRequest.getPageNum();
        long pageSize = userQueryRequest.getPageSize();
        Page<User> userPage = userService.page(Page.of(pageNum, pageSize),
                userService.getQueryWrapper(userQueryRequest));
        // 数据脱敏
        Page<UserVO> userVOPage = new Page<>(pageNum, pageSize, userPage.getTotalRow());
        List<UserVO> userVOList = userService.getUserVOList(userPage.getRecords());
        userVOPage.setRecords(userVOList);
        return ResultUtils.success(userVOPage);
    }

    /**
     * 修改密码
     *
     * @param userUpdatePasswordRequest 修改密码请求
     * @param request                   请求对象
     * @return 修改结果
     */
    @PostMapping("/update/password")
    public BaseResponse<Boolean> updatePassword(@RequestBody UserUpdatePasswordRequest userUpdatePasswordRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(userUpdatePasswordRequest == null, ErrorCode.PARAMS_ERROR);
        User loginUser = userService.getLoginUser(request);
        boolean result = userService.updatePassword(
                userUpdatePasswordRequest.getOldPassword(),
                userUpdatePasswordRequest.getNewPassword(),
                userUpdatePasswordRequest.getCheckPassword(),
                loginUser.getId()
        );
        return ResultUtils.success(result);
    }

    /**
     * 修改用户信息
     *
     * @param userUpdateInfoRequest 修改用户信息请求
     * @param request               请求对象
     * @return 修改结果
     */
    @PostMapping("/update/info")
    public BaseResponse<Boolean> updateUserInfo(@RequestBody UserUpdateInfoRequest userUpdateInfoRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(userUpdateInfoRequest == null, ErrorCode.PARAMS_ERROR);
        User loginUser = userService.getLoginUser(request);
        boolean result = userService.updateUserInfo(
                userUpdateInfoRequest.getUserName(),
                userUpdateInfoRequest.getUserProfile(),
                loginUser.getId()
        );
        return ResultUtils.success(result);
    }

    /**
     * 修改用户头像
     *
     * @param request    请求对象
     * @return 修改结果
     */
    @PostMapping("/update/avatar")
    public BaseResponse<Boolean> updateUserAvatar(@RequestBody UserUpdateAvatarRequest userUpdateAvatarRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(userUpdateAvatarRequest == null || StrUtil.isBlank(userUpdateAvatarRequest.getUserAvatar()), ErrorCode.PARAMS_ERROR);
        User loginUser = userService.getLoginUser(request);
        boolean result = userService.updateUserAvatar(userUpdateAvatarRequest.getUserAvatar(), loginUser.getId());
        return ResultUtils.success(result);
    }
}
